internal/pkg/mqtt/client_options.go

```diff
@@ -0,0 +1,238 @@
+/********************************************************************************
+ *  Copyright 2020 Dell Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ *******************************************************************************/
+
+package mqtt
+
+import (
+	"net/url"
+	"strconv"
+	"time"
+
+	mqtt "github.com/eclipse/paho.mqtt.golang"
+
+	"github.com/edgexfoundry/go-mod-messaging/pkg/types"
+)
+
+const (
+	// Names of the optional configuration properties as they are expected to appear in the MessageBusConfig's Optional
+	// field.
+	ClientID             = "client-id"
+	Username             = "username"
+	Password             = "password"
+	CleanSession         = "CleanSession"
+	Order                = "Order"
+	WillEnabled          = "WillEnabled"
+	WillTopic            = "WillTopic"
+	WillPayload          = "WillPayload"
+	WillQos              = "WillQos"
+	WillRetained         = "WillRetained"
+	ProtocolVersion      = "ProtocolVersion"
+	AutoReconnect        = "AutoReconnect"
+	ResumeSubs           = "ResumeSubs"
+	KeepAlive            = "KeepAlive"
+	PingTimeout          = "PingTimeout"
+	ConnectTimeout       = "ConnectTimeout"
+	MaxReconnectInterval = "MaxReconnectInterval"
+	WriteTimeout         = "WriteTimeout"
+	MessageChannelDepth  = "MessageChannelDepth"
+)
+
+// MandatoryClientOptions defines the configuration properties which are required.
+var MandatoryClientOptions = []string{
+	// The ClientID is needed to avoid connection issues where the MQTT will not establish a connection if not
+	// specified.
+	ClientID,
+}
+
+// toClientOptions creates mqtt.ClientOptions based on the information available in the provided options.
+//
+// NOTE: The following 'mqtt.ClientOptions' are not propagated to the underlying mqtt.ClientOptions:
+// - HTTPHeaders
+// - CredentialsProvider
+// - TLSConfig
+// - Store
+// - DefaultPublishHandler
+// - OnConnect
+// - OnConnectionLost
+func toClientOptions(options types.MessageBusConfig, defaults mqtt.ClientOptions) (mqtt.ClientOptions, error) {
```

Comment:
Checkout how the MQTT Device service handles transferring the Map values to a config structure using reflection. 

This package does what we want, but with `map[string]interface{}`.
https://godoc.org/github.com/mitchellh/mapstructure

Not sure if TOML and/or Consul can handle `map[string]interface{}` instead of `map[string]string`